### ISSUE TITLE
Validator with pex=false on a connection lost is dialing the internal nat IP  instead of external ip setup in  persistent_peers

### ISSUE THREAD

**Issue by CzMcClane** [tendermint/tendermint#1720]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):BUG REPORT

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):
0.19.7-dev (seen it in older versions also)

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
GaiaApp(Cosmos)
**Environment**:
- **OS** (e.g. from /etc/os-release): Centos 7
- **Install tools**:
- **Others**:

**What happened**:
if a validator with pex=false is loosing the connection to one of the sentry that is behind nat, instead of dialing the external ip of the sentry that is setup into the persistent_peers, it is dialing the internal ip of the sentry, the one that is shown in listen_addr. 

**What you expected to happen**:
The validator should keep trying to dial the external ip setup in persistent_peers until the sentry gets back online

**How to reproduce it** (as minimally and precisely as possible):
this can be easily replicated buy blocking the access from the validator to the sentry and watching the validator trying to dial the internal ip and not recover.  run route add sentry_ip dev lo and watch the logs of the validator, after run route del sentry_ip dev lo and you will see that the validator will keep trying to connect to internal ip

**Logs (you can paste a part showing an error or attach the whole file)**:

```
I[06-08|17:04:29.413] Added peer                                   module=p2p peer="Peer{MConn{**35.180.98.65**:46656} c7439aeb9a4142e9a61b5ad9ec1685cd884fb784 out}"
E[06-08|17:32:14.414] Connection failed @ sendRoutine              module=p2p peer=35.180.98.65:46656 conn=MConn{35.180.98.65:46656} err="pong timeout"
I[06-08|17:32:14.414] Stopping MConnection                         module=p2p peer=35.180.98.65:46656 impl=MConn{35.180.98.65:46656}
E[06-08|17:32:14.414] Stopping peer for error                      module=p2p peer="Peer{MConn{35.180.98.65:46656} c7439aeb9a4142e9a61b5ad9ec1685cd884fb784 out}" err="pong timeout"
I[06-08|17:32:14.414] Stopping Peer                                module=p2p peer=35.180.98.65:46656 impl="Peer{MConn{35.180.98.65:46656} c7439aeb9a4142e9a61b5ad9ec1685cd884fb784 out}"
I[06-08|17:32:14.414] Reconnecting to peer                         module=p2p addr=c7439aeb9a4142e9a61b5ad9ec1685cd884fb784@**172.31.37.117**:46656
I[06-08|17:32:14.415] Dialing peer                                 module=p2p address=c7439aeb9a4142e9a61b5ad9ec1685cd884fb784@172.31.37.117:46656
I[06-08|17:32:14.457] Stopping gossipDataRoutine for peer          module=consensus peer="Peer{MConn{35.180.98.65:46656} c7439aeb9a4142e9a61b5ad9ec1685cd884fb784 out}"
I[06-08|17:32:14.457] Stopping gossipVotesRoutine for peer         module=consensus peer="Peer{MConn{35.180.98.65:46656} c7439aeb9a4142e9a61b5ad9ec1685cd884fb784 out}"
I[06-08|17:32:15.635] Stopping queryMaj23Routine for peer          module=consensus peer="Peer{MConn{35.180.98.65:46656} c7439aeb9a4142e9a61b5ad9ec1685cd884fb784 out}"
I[06-08|17:32:17.415] Error reconnecting to peer. Trying again     module=p2p tries=0 err="Error{`Error creating peer` (cause: dial tcp 172.31.37.117:46656: i/o timeout)}" addr=c7439aeb9a4142e9a61b5ad9ec1685cd884fb784@172.31.37.117:46656
I[06-08|17:32:25.349] Dialing peer                                 module=p2p address=c7439aeb9a4142e9a61b5ad9ec1685cd884fb784@172.31.37.117:46656
I[06-08|17:32:28.350] Error reconnecting to peer. Trying again     module=p2p tries=1 err="Error{`Error creating peer` (cause: dial tcp 172.31.37.117:46656: i/o timeout)}" addr=c7439aeb9a4142e9a61b5ad9ec1685cd884fb784@172.31.37.117:46656
```

**`/dump_consensus_state` output for consensus bugs**

**Anything else do we need to know**:

**Comment by xla** [comment]
Related #873

**Comment by awrelll** [comment]
I think i'm also having the same type of situation. I keep being revoked :(
```
E[06-23|10:38:34.460] Connection failed @ recvRoutine (reading byte) module=p2p peer=13.58.19.0:44468 conn=MConn{13.58.19.0:44468} err=EOF
I[06-23|10:38:34.461] Stopping MConnection                         module=p2p peer=13.58.19.0:44468 impl=MConn{13.58.19.0:44468}
E[06-23|10:38:34.461] Stopping peer for error                      module=p2p peer="Peer{MConn{13.58.19.0:44468} 7127568cf9885ab50c780bb7762e64cccf4e13e9 in}" err=EOF
I[06-23|10:38:34.461] Stopping Peer                                module=p2p peer=13.58.19.0:44468 impl="Peer{MConn{13.58.19.0:44468} 7127568cf9885ab50c780bb7762e64cccf4e13e9 in}"
E[06-23|10:38:34.461] MConnection flush failed                     module=p2p peer=13.58.19.0:44468 err="write tcp 45.77.53.208:46656->13.58.19.0:44468: use of closed network connection"
```

**Comment by melekes** [comment]
0.22.0 release comes with `p2p.external_address` config option. Does it solves your problem?

```
# Address to advertise to peers for them to dial
# If empty, will use the same port as the laddr,
# and will introspect on the listener or use UPnP
# to figure out the address.
external_address = ""
```

**Comment by ebuchman** [comment]
external_address helps but we should still redial the original address we dialed, not the address we got from the peer during the handshake.

**Comment by xla** [comment]
@melekes Can we close this one with #1994 being merged?

**Comment by melekes** [comment]
Merged to develop. Will be shipped with 0.22.5 or next breaking release.
